feat: WIP - Create change requests for segments

zachaysan opened the pull request:
## Changes

I'm working through creating change requests for segments on both this repository and on the workflows one. I'm going to try to keep WIP PRs to avoid losing my work if there is data loss.

## How did you test this code?

Tests still required, although most of it will eventually live in the workflows repository.

matthewelwell commented on api/features/workflows/core/models.py:
```diff
@@ -179,6 +180,30 @@ def _publish_change_sets(self, published_by: "FFAdminUser") -> None:
             change_set.publish(user=published_by)
 
+    def _publish_segments(self) -> None:
+        segments = list(self.segments.all())
+        if not segments:
+            return
+
+        for segment in segments:
```
Is there a reason to do the empty check separately instead of just relying on the `for` loop? e.g. 

```suggestion
        for segment in self.segments.all():
```

zachaysan replied:
Nice catch, implemented

matthewelwell commented on api/features/workflows/core/models.py:
```diff
@@ -179,6 +180,30 @@ def _publish_change_sets(self, published_by: "FFAdminUser") -> None:
+            assert target_segment != segment
+
+            # Deep clone the segment to establish historical version.
+            target_segment.deep_clone()
```
Couldn't we optimise this by sending in an argument, or via a separate method, which doesn't clone the target_segments rules? Since we're just deleting them again below, it seems unnecessary to do this.

zachaysan replied:
This is for establishing a versioned history for the soon-to-be-changing segment. Without a deep clone we won't have a copy of a previous version of the segment with the associated rules and conditions. We can always skip doing the deep clone but then our version history would have gaps in it if change requests are used to change segments.

matthewelwell replied:
I don't think that's true is it? I get that we want to keep the history, but we don't need to actually duplicate the rules and conditions, we just need to archive the ones from the current (outgoing) segment. We don't need to create them for the new segment because we're not going to use them - we just delete them below.

matthewelwell replied:
As discussed, this should be thought of like regular updates to a segment where we create the clone to create the version, then modifying the new 'draft' version with the data from the change request. 

Let's update the comments for this method to try and make this a little clearer.

zachaysan replied:
Ok the comment has been updated.

matthewelwell commented on api/projects/models.py:
```diff
@@ -142,19 +142,22 @@ def edge_v2_identity_overrides_migrated(self) -> bool:
         return self.edge_v2_migration_status == EdgeV2MigrationStatus.COMPLETE
 
     def get_segments_from_cache(self):
+        from segments.models import Segment
```
Hmm... can we avoid this if we convert this method into a function and move it out of `models`? Since this method gets called in some pretty hot paths, I'd rather avoid local imports.

zachaysan replied:
I'll move it to a function, but I think we should keep the method on project even if it's just a reference to the function because it's a natural calling path.

zachaysan replied:
Ok so I moved it to the function but I'm calling `apps.get_model("segments", "Segment")` in the function because of the circular import issue. Is that ok or does that still trigger your hot path sensitivity?

matthewelwell replied:
Hmm... interesting dilemma, I'm not sure. I think I prefer not using `apps.get_model()` as I suspect that also involves a local import. 

I'm keen to get other opinions on this - @gagantrivedi

zachaysan replied:
I believe the models are stored in a [registry](https://docs.djangoproject.com/en/5.1/ref/applications/#django.apps.AppConfig.get_model) that can be looked up quickly during a call to `apps.get_model()` so I think we are safe in using this here. It makes sense that Django would cache the models in, say a list or dictionary.

matthewelwell commented on api/segments/models.py:
```diff
@@ -60,24 +59,30 @@ class Segment(
         blank=True,
     )
+
+    change_request = models.ForeignKey(
+        "workflows_core.ChangeRequest",
+        on_delete=models.CASCADE,
+        null=True,
+        related_name="segments",
+    )
```
let's add `blank=True` here too in case it appears in the django admin anywhere

zachaysan replied:
Good idea. Implemented.

matthewelwell commented on api/segments/models.py:
```diff
@@ -145,6 +150,26 @@ def set_version_of_to_self_if_none(self):
         self.save()
         segment_audit_log_helper.unset_skip_audit_log(self.id)
 
+    def shallow_clone(
```
Ah, is this the method I described above?

matthewelwell commented on api/tests/unit/features/workflows/core/test_unit_workflows_models.py:
```diff
@@ -773,3 +795,61 @@ def test_change_request_live_from_for_change_request_with_change_set(
+    )
+
+    # When
+    change_request._publish_segments()
```
Why not just call `change_request.commit()` ?

zachaysan replied:
Sure that's no problem

matthewelwell commented on api/tests/unit/projects/test_unit_projects_models.py:
```diff
@@ -55,6 +56,35 @@ def test_get_segments_from_cache_set_not_called(project, segments, monkeypatch):
     mock_project_segments_cache.set.assert_not_called()
 
 
+@pytest.mark.django_db()
```
We don't need this since the `project` and `segment` fixtures will require the `db` fixture already.

matthewelwell commented on api/util/mappers/engine.py:
```diff
@@ -194,7 +195,17 @@ def map_environment_to_engine(
     project_segments: List["Segment"] = project.segments.all()
+
+    # Even though the main calling code filters via a prefetch,
+    # we still want to gaurd this function in case there are other
```
```suggestion
    # we still want to guard this function in case there are other
```

🎃

matthewelwell reviewed (commented):
In general this looks really good. I've added a few minor comments. 

One other thing that stood out to me was around audit trail - are we happy that we're writing the correct audit logs here? I guess we're covered by those written for the change request itself?

zachaysan commented:
> In general this looks really good. I've added a few minor comments.
> 
> One other thing that stood out to me was around audit trail - are we happy that we're writing the correct audit logs here? I guess we're covered by those written for the change request itself?

The audit trail tests have already been implemented when working on the segments versioned code. In addition to that, the audit trail logs are implemented in the workflows repository.

matthewelwell commented on api/projects/migrations/0025_add_change_request_project_permissions.py:
```diff
@@ -0,0 +1,61 @@
+
+def insert_default_project_permissions(apps, schema_model):
+    PermissionModel = apps.get_model("permissions", "PermissionModel")
+    manage_description = approve_description = None
+
+    for project_permission in PROJECT_PERMISSIONS:
+        if project_permission[0] == MANAGE_PROJECT_LEVEL_CHANGE_REQUESTS:
+            manage_description = project_permission[1]
+        elif project_permission[0] == APPROVE_PROJECT_LEVEL_CHANGE_REQUESTS:
+            approve_description = project_permission[1]
+
+    assert manage_description
+    assert approve_description
```
This code took me a little while to understand. I don't think we've worried about doing this dynamically in the past. I don't mind having the duplication here really. 

Alternatively, if you really want to keep this, maybe let's add a comment. Or, just define the descriptions as constants alongside the permissions themselves.

zachaysan replied:
Ok I've duplicated the descriptions directly in the code

matthewelwell commented on api/projects/models.py:
```diff
@@ -106,6 +106,7 @@ class Project(LifecycleModelMixin, SoftDeleteExportableModel):
         default=30,
         help_text="Number of days without modification in any environment before a flag is considered stale.",
     )
+    minimum_change_request_approvals = models.IntegerField(blank=True, null=True)
```
This will also need adding to relevant serializer(s).

zachaysan replied:
Alright I've added it to `ProjectListSerializer` which looks like it will include the field to `ProjectRetrieveSerializer` and other serializers.

matthewelwell commented on api/tests/unit/features/workflows/core/test_unit_workflows_models.py:
```diff
@@ -897,3 +977,31 @@ def _create_segment(percentage_value: int) -> Segment:
+def test_approval_via_project(project_change_request: ChangeRequest) -> None:
+    # Given - The project change request fixture
+    assert project_change_request.environment is None
+    assert project_change_request.project.minimum_change_request_approvals is None
+
+    # When
+    is_approved = project_change_request.is_approved()
+
+    # Then
+    assert is_approved is True
```
Is this a legit use case? For environments at least, `minimum_change_request_approvals = None` means that change requests aren't enabled at all.

zachaysan replied:
According to a pre-existing test (`test_change_request_is_approved_returns_true_when_minimum_change_request_approvals_is_none`) I believe your understanding is incorrect. I just mirrored the behavior from the environments definition for the project piece.

matthewelwell commented on api/tests/unit/features/workflows/core/test_unit_workflows_models.py:
```diff
@@ -897,3 +977,31 @@ def _create_segment(percentage_value: int) -> Segment:
+    url = project_change_request.url
+
+    # Then
+    project_id = project_change_request.project.id
```
... feel bad nitpicking after all the other comments, but this is a particularly bad habit to get into with django, so I'm calling it out here, even though it's just test code. 

```suggestion
    project_id = project_change_request.project_id
```

zachaysan replied:
No problem, I've updated the code to match

zachaysan commented on api/app/urls.py:
```diff
@@ -54,8 +55,24 @@
 
 if settings.WORKFLOWS_LOGIC_INSTALLED:
     workflow_views = importlib.import_module("workflows_logic.views")
+    from projects.views import ProjectViewSet  # pragma: no cover
```
Ok I've moved the urls to the projects urls module, but for some reason I'm dealing with a non-update of the image when I try to push the code so I haven't been able to test the change yet.

zachaysan replied:
Ok it looks like the images are being created again. Must have either been a merge conflict with `main` or an intermittent outage. I was able to use the new urls after requiring to update the namespacing.

zachaysan commented on api/features/workflows/core/migrations/0011_add_project_to_change_requests.py:
```diff
@@ -0,0 +1,62 @@
+        environment_id__isnull=False
+    ).select_related("environment"):
+        change_request.project = change_request.environment.project  # pragma: no cover
+        change_request.save()  # pragma: no cover
```
Good idea. I've written the migration test.

zachaysan commented on api/features/workflows/core/models.py:
```diff
@@ -77,10 +77,19 @@ class ChangeRequest(
+        "projects.Project",
+        on_delete=models.SET_NULL,
+        related_name="change_requests",
+        null=True,
```
Sounds good. I've done these changes.

zachaysan commented on api/features/workflows/core/models.py:
```diff
@@ -209,9 +243,24 @@ def _get_environment(self) -> typing.Optional["Environment"]:
+            return self.is_approved_via_environment()
+        if self.project:
+            return self.is_approved_via_project()
+        raise RuntimeError(  # pragma: no cover
+            "Unable to approve change request without environment or project"
+        )
```
Yep. Agreed. Have removed the exception.

zachaysan commented on api/features/workflows/core/models.py:
```diff
@@ -228,8 +277,15 @@ def url(self):
         url = get_current_site_url()
-        url += f"/project/{self.environment.project_id}"
-        url += f"/environment/{self.environment.api_key}"
+        if self.environment:
+            url += f"/project/{self.environment.project_id}"
+            url += f"/environment/{self.environment.api_key}"
+        elif self.project_id:
+            url += f"/projects/{self.project_id}"
+        else:
+            raise RuntimeError(  # pragma: no cover
+                "Change request missing fields for URL"
+            )
```
Yep. Simplified it.

zachaysan commented on api/features/workflows/core/tasks.py:
```diff
@@ -0,0 +1,8 @@
+
+@register_task_handler()
+def delete_change_request(change_request_id: int) -> None:
+    ChangeRequest.objects.get(pk=change_request_id).delete()
```
Since project is now non-nullable I switched this to just use the cascading deletion by default.

matthewelwell commented on api/features/workflows/core/migrations/0011_add_project_to_change_requests.py:
```diff
@@ -0,0 +1,72 @@
+
+    for change_request in ChangeRequest.objects.filter(
+        environment_id__isnull=False
+    ).select_related("environment"):
```
```suggestion
    ).select_related("environment", "environment__project"):
```

matthewelwell reviewed (commented):
Added a couple of minor comments, but this looks good on the whole. I guess the only question I have is whether there should also be a bunch of removed code here based on the stuff we moved to the flagsmith-common repo?

zachaysan commented on api/features/workflows/core/migrations/0011_add_project_to_change_requests.py:
```diff
@@ -0,0 +1,72 @@
+        environment_id__isnull=False
+    ).select_related("environment"):
+        change_request.project = change_request.environment.project
+        change_request.save()
```
Yep! Implemented.

zachaysan commented on api/features/workflows/core/models.py:
```diff
@@ -77,10 +78,19 @@ class ChangeRequest(
         null=True,
     )
 
+    # Change requests get deleted in a delegated task when a project is deleted.
```
Good catch! Removed.

zachaysan commented:
> Added a couple of minor comments, but this looks good on the whole. I guess the only question I have is whether there should also be a bunch of removed code here based on the stuff we moved to the flagsmith-common repo?

Ok it took a while, but I've updated the repo with references to the flagsmith-common repo and of course removed the now unnecessary code.

zachaysan commented on api/permissions/migrations/0001_initial.py:
Ok thanks for letting me know, although I'm pretty sure the formatting changes here are fine.

zachaysan commented on api/segments/views.py:
```diff
@@ -39,7 +40,7 @@ def get_queryset(self):
         )
         project = get_object_or_404(permitted_projects, pk=self.kwargs["project_pk"])
 
-        queryset = project.segments.all()
+        queryset = Segment.live_objects.filter(project=project)
```
I believe some were missed or maybe there was a refactor in this PR where calling sites had to be upgraded.

zachaysan replied:
Yes it was a refactor, see my comment below.

zachaysan commented on api/tests/unit/segments/test_unit_segments_views.py:
```diff
@@ -619,7 +619,7 @@ def test_update_segment_versioned_segment(
     # Before updating the segment confirm pre-existing version count which is
     # automatically set by the fixture.
-    assert Segment.all_objects.filter(version_of=segment).count() == 2
+    assert Segment.objects.filter(version_of=segment).count() == 2
```
This PR changed the way objects are accessed. As you can see [here](https://github.com/Flagsmith/flagsmith/pull/4265/files#diff-13990717d51e5dbadd003d4a2fe9bf926d5238232e80aa73c48f450e4b02719bR86)

matthewelwell replied:
Hmm... in that case, I want to be really sure that we've updated ALL usages. For example, I think there's still a few that might need updating? Some examples:

 * https://github.com/Flagsmith/flagsmith/blob/main/api/integrations/launch_darkly/services.py#L244
 * https://github.com/Flagsmith/flagsmith/blob/main/api/util/mappers/engine.py#L197
 * https://github.com/Flagsmith/flagsmith/blob/main/api/projects/models.py#L122
 * https://github.com/Flagsmith/flagsmith/blob/main/api/projects/models.py#L151

There are a few more that I found using this regex search: `((?<!\w)Segment\.objects|(?<!feature_)segments)\.(get|filter|all|count)` which should find all instances of `Segment.objects.get|filter` and `x.segments.all|filter|count`. 

I ignored the ones found in the test directory, but all others should be looked at.

zachaysan replied:
Ok these are updated and I've pushed the fix to the conflicts below.

matthewelwell reviewed (commented):
This all looks good to me now - there's just the (new?) conflicts to resolve, and the CI failures.

zachaysan commented:
> This all looks good to me now - there's just the (new?) conflicts to resolve, and the CI failures.

Conflicts and CI failures have been resolved. Are we good to merge (other than the update to the workflows repo which needs to be merged itself)?

matthewelwell reviewed (approved):
I'm approving this now, but we need to consider the release plan as this is quite a meaty change. It will need a fair amount of regression testing in staging.

khvn26 commented on api/projects/migrations/0025_add_change_request_project_permissions.py:
```diff
@@ -0,0 +1,48 @@
+    manage_description = "Ability to manage change requests associated with a project."
+    approve_description = "Ability to approve project level change requests."
+
+    PermissionModel.objects.get_or_create(
+        key=MANAGE_PROJECT_LEVEL_CHANGE_REQUESTS,
+        description=manage_description,
+        type=PROJECT_PERMISSION_TYPE,
+    )
+    PermissionModel.objects.get_or_create(
+        key=APPROVE_PROJECT_LEVEL_CHANGE_REQUESTS,
+        description=approve_description,
+        type=PROJECT_PERMISSION_TYPE,
+    )
```
```python
    PermissionModel.objects.get_or_create(
        key=MANAGE_PROJECT_LEVEL_CHANGE_REQUESTS,
        type=PROJECT_PERMISSION_TYPE,
        defaults={"description": manage_description},
    )
    PermissionModel.objects.get_or_create(
        key=APPROVE_PROJECT_LEVEL_CHANGE_REQUESTS,
        type=PROJECT_PERMISSION_TYPE,
        defaults={"description": manage_description},
    )
```
